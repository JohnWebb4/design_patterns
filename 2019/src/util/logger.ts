/* tslint:disable:no-console */

class Logger {
  public static instance() {
    return this.logger;
  }

  private static logger = new Logger();

  private constructor() {}

  public log(...params: any[]) {
    console.log(...params);
  }
}

const logger = Logger.instance();

export { logger };
