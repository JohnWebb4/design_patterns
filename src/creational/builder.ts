import { logger } from "../util/logger";

interface IPrepare {
  prepare: () => string;
  prepareFee: number;
}

interface IFood {
  cost: number;
  name: string;
  prepare: () => IPrepare;
}

class Cookable implements IPrepare {
  public prepareFee = 3;

  public prepare() {
    return "Cooked";
  }
}

class Serveable implements IPrepare {
  public prepareFee = 1;

  public prepare() {
    return "Served";
  }
}

class Burger implements IFood {
  public cost = 4;

  public name = "Burger";

  public prepare() {
    return new Cookable();
  }
}

class FrenchFries implements IFood {
  public cost = 2;

  public name = "French Fries";

  public prepare() {
    return new Serveable();
  }
}

class DoubleBurger extends Burger {
  public cost = 6;

  public name = "Double Burger";
}

class EverythingBurger extends Burger {
  public cost = 5;

  public name = "Everything Burger";
}

class CurlyFries extends FrenchFries {
  public cost = 3;

  public name = "Curly Fries";
}

class CajunFries extends FrenchFries {
  public cost = 3;

  public name = "Cajun Fries";
}

class Meal {
  private food: IFood[] = [];

  public orderFood(food: IFood) {
    return this.food.push(food);
  }

  public getPrice() {
    return this.food.reduce((totalPrice, food) => totalPrice + food.cost, 0);
  }

  public listMeal() {
    this.food.forEach((food) => {
      logger.log(`Food: ${food.name}, Cost: ${food.cost}`);
    });
  }
}

class MealBuilder {
  public prepareDoubleWonderMeal(): Meal {
    const meal = new Meal();

    meal.orderFood(new DoubleBurger());
    meal.orderFood(new CurlyFries());

    return meal;
  }

  public prepareSpicyMeal(): Meal {
    const meal = new Meal();

    meal.orderFood(new EverythingBurger());
    meal.orderFood(new CajunFries());

    return meal;
  }
}

export { CajunFries, CurlyFries, DoubleBurger, EverythingBurger, Meal, MealBuilder };
