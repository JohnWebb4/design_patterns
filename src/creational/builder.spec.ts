import { logger } from "../util/logger";
import { CajunFries, EverythingBurger, Meal, MealBuilder } from "./builder";

jest.mock("../util/logger", () => ({
  logger: {
    log: jest.fn(),
  },
}));

describe("Builder", () => {

  describe("Meal Builder", () => {
    let mealBuilder: MealBuilder;

    beforeEach(() => {
      mealBuilder = new MealBuilder();
    });
    it("should create a double wonder meal", () => {
      const meal = mealBuilder.prepareDoubleWonderMeal();

      expect(meal).toMatchSnapshot();
    });

    it("should create spicy meal", () => {
      const meal = mealBuilder.prepareSpicyMeal();

      expect(meal).toMatchSnapshot();
    });

    it("should add food to order", () => {
      const meal = mealBuilder.prepareDoubleWonderMeal();

      meal.orderFood(new CajunFries());

      expect(meal).toMatchSnapshot();
    });
  });

  describe("Cajun Fries", () => {
    let cajunFries: CajunFries;

    beforeEach(() => {
      cajunFries = new CajunFries();
    });

    describe("when preparing meal", () => {
      it("should return preparable", () => {
        expect(cajunFries.prepare()).toMatchSnapshot();
      });

      it("should return served when called", () => {
        expect(cajunFries.prepare().prepare()).toEqual("Served");
      });
    });

  });

  describe("Everything Burger", () => {
    let everythingBurger: EverythingBurger;

    beforeEach(() => {
      everythingBurger = new EverythingBurger();
    });

    describe("when preparing meal", () => {
      it("should return preparable", () => {
        expect(everythingBurger.prepare()).toMatchSnapshot();
      });

      it("should return cooked when called", () => {
        expect(everythingBurger.prepare().prepare()).toEqual("Cooked");
      });
    });
  });

  describe("Meal", () => {
    let meal: Meal;
    let mealBuilder: MealBuilder;

    beforeEach(() => {
      mealBuilder = new MealBuilder();
      meal = mealBuilder.prepareDoubleWonderMeal();
    });

    describe("Get Price", () => {
      it("should sum items for meal price", () => {
        expect(meal.getPrice()).toEqual(9);
      });
    });

    describe("List items", () => {
      it("should list items", () => {
        meal.listMeal();

        expect(logger.log).toHaveBeenCalledWith("Food: Curly Fries, Cost: 3");
        expect(logger.log).toHaveBeenCalledWith("Food: Double Burger, Cost: 6");
      });
    });
  });
});
