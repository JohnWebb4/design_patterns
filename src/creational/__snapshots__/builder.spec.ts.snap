// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder Cajun Fries when preparing meal should return preparable 1`] = `
Serveable {
  "prepareFee": 1,
}
`;

exports[`Builder Everything Burger when preparing meal should return preparable 1`] = `
Cookable {
  "prepareFee": 3,
}
`;

exports[`Builder Meal Builder should add food to order 1`] = `
Meal {
  "food": Array [
    DoubleBurger {
      "cost": 6,
      "name": "Double Burger",
    },
    CurlyFries {
      "cost": 3,
      "name": "Curly Fries",
    },
    CajunFries {
      "cost": 3,
      "name": "Cajun Fries",
    },
  ],
}
`;

exports[`Builder Meal Builder should create a double wonder meal 1`] = `
Meal {
  "food": Array [
    DoubleBurger {
      "cost": 6,
      "name": "Double Burger",
    },
    CurlyFries {
      "cost": 3,
      "name": "Curly Fries",
    },
  ],
}
`;

exports[`Builder Meal Builder should create spicy meal 1`] = `
Meal {
  "food": Array [
    EverythingBurger {
      "cost": 5,
      "name": "Everything Burger",
    },
    CajunFries {
      "cost": 3,
      "name": "Cajun Fries",
    },
  ],
}
`;
